docs/design.md

```diff
@@ -0,0 +1,299 @@
+# Design Doc: Piece Indexer
+
+## Introduction
+
+### Context
+
+Filecoin defines a `Piece` as the main unit of negotiation for data that users
+store on the Filecoin network. This is reflected in the on-chain metadata field
+`PieceCID`.
+
+On the other hand, content is retrieved from Filecoin using the CID of the
+requested payload.
+
+This dichotomy poses a challenge for retrieval checkers like Spark: for a given
+deal storing some PieceCID, what payload CID request when testing the retrieval?
+
+Spark v1 relies on StorageMarket's DealProposal metadata `Label`, which is often
+(but not always!) set by the client to the root CID of the payload stored.
+
+In the first half of 2024, the Filecoin network added support for DDO (Direct
+Data Onboarding) deals. These deals don't have any DealProposal, there is no
+`Label` field. Only `PieceCID`. We need to find a different solution to support
+these deals.
+
+In the longer term, we want IPNI to provide a reverse lookup functionality that
+will allow clients like Spark to request a sample of payload block CIDs
+associated with the given ContextID or PieceCID, see the
+[design proposal for InterPlanetary Piece Index](https://docs.google.com/document/d/1jhvP48ccUltmCr4xmquTnbwfTSD7LbO1i1OVil04T2w).
+
+To cover the gap until the InterPlanetary Piece Index is live, we want to
+implement a lightweight solution that can serve as a stepping stone between
+Label-based CID discovery and full Piece Index sampling.
+
+### High-Level Idea
+
+Let's implement a lightweight IPNI ingester that will process the advertisements
+from Filecoin SPs and extract the list of `(ProviderID, PieceCID, PayloadCID)`
+entries. Store these entries in a Postgres database. Provide a REST API endpoint
+accepting `(ProviderID, PieceCID)` and returning a single `PayloadCID`.
+
+#### Notes
+
+Pieces are immutable. If we receive an advertisement saying that a payload block
+CID was found in a piece CID, then this information remains valid forever, even
+after the SP advertise that they are no longer storing that block. This means
```

Comment:
Yes, we should not make any retrievals for such payload.

I am envisioning the following architecture:

1. A piece indexer I describe in this document. It will be eventually replaced by a proper IPNI reverse index solution.
2. A deal tracker - a component that listens for actor events and builds a list of active deals - conceptually a list of pairs `(piece_cid, miner_id)`. We will need to build this. 

When Spark build a list of tasks for the current round, it will ask the deal tracker for 1000 active deals. This ensures we test retrieval for active deals only.

When Spark checker tests retrieval, it will first consult the piece indexer to convert deal's `piece_cid` to a payload CID to retrieve.

It's ok if the piece indexer stores data for expired deals, because Spark is not going to ask for that data. 

Of course, storing expired deals unnecessarily increases storage requirements, but since we want to run this service only for 2-6 months, I think it's fine.